prune/archivewriter.go

```diff
@@ -0,0 +1,166 @@
+package prune
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog/log"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	_ "gocloud.dev/blob/gcsblob"
+	_ "gocloud.dev/blob/s3blob"
+)
+
+// Bucket defines the interface for cloud storage operations.
+type Bucket interface {
+	NewReader(ctx context.Context, key string, opts *blob.ReaderOptions) (Reader, error)
+	NewWriter(ctx context.Context, key string, opts *blob.WriterOptions) (Writer, error)
+	Copy(ctx context.Context, dstKey, srcKey string, opts *blob.CopyOptions) error
+	Exists(ctx context.Context, key string) (bool, error)
+}
+
+// Reader defines the interface for reading from cloud storage objects.
+type Reader interface {
+	io.ReadCloser
+	Size() int64
+}
+
+// Writer defines the interface for writing to cloud storage objects.
+type Writer interface {
+	io.WriteCloser
+}
+
+// blobBucket implements the Bucket interface using "gocloud.dev/blob".
+type blobBucket struct {
+	*blob.Bucket
+}
+
+// NewReader creates a new Reader for the given object key.
+func (b *blobBucket) NewReader(ctx context.Context, key string, opts *blob.ReaderOptions) (Reader, error) {
+	return b.Bucket.NewReader(ctx, key, opts)
+}
+
+// NewWriter creates a new Writer for the given object key.
+func (b *blobBucket) NewWriter(ctx context.Context, key string, opts *blob.WriterOptions) (Writer, error) {
+	return b.Bucket.NewWriter(ctx, key, opts)
+}
+
+// NewBlobBucket creates a new Bucket using "gocloud.dev/blob".
+func NewBlobBucket(bucket *blob.Bucket) Bucket {
+	return &blobBucket{bucket}
+}
+
+// ArchiveWriteManager handles file rotation for cloud storage.
+type ArchiveWriteManager struct {
+	bucket      Bucket
+	objectName  string
+	maxSize     int64
+	currentSize int64
+	mu          sync.Mutex
+	writer      Writer
+}
+
+// NewArchiveWriteManager creates a new RotationManager.
+func NewArchiveWriteManager(bucket Bucket, objectName string, maxSize int64) (*ArchiveWriteManager, error) {
+	rm := &ArchiveWriteManager{
+		bucket:     bucket,
+		objectName: objectName,
+		maxSize:    maxSize,
+	}
+
+	// Initialize currentSize using reader
+	ctx := context.Background()
+	r, err := bucket.NewReader(ctx, objectName, nil)
+	exists, exists_err := bucket.Exists(ctx, objectName)
+	if exists_err != nil {
+		return nil, fmt.Errorf("failed to check if object exists: %w", exists_err)
+	}
+	if err != nil && exists {
+		return nil, fmt.Errorf("failed to create reader: %w", err)
+	}
+	if err == nil {
+		defer r.Close()
+		rm.currentSize = r.Size()
+	}
```

Comment:
Did we intentionally ignored a case here? `err != nil && !exists`?

Reply:
Yes, we ignore the open error if the file does not exist, IOW file needs to be created